### ISSUE TITLE
Add a parameter to Read-Host to allow prefilling the edit buffer

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#14013]
## Summary of the new feature/enhancement

It would improve the end-user experience if a `Read-Host` call were allowed to _pre-fill the edit buffer_ with a default value that the user can either accept as-is or modify:

```powershell
# WISHFUL THINKING
$var = Read-Host -Prompt 'Enter a path' -DefaultValue '/path/to/default/location'
```

The user would then see:

```none
Enter a path: /path/to/default/location
```

with the cursor at the end of the value, allowing them to press <kbd>Enter</kbd> to accept the default value as-is, or modify the value as needed.

## Proposed technical implementation details

If `PSReadLine`'s `PSConsoleHostReadLine` function were to be enhanced to become a general-purpose prompting function, as proposed in [issue #881](https://github.com/PowerShell/PSReadLine/issues/881), the implementation of `-DefaultValue` could build on that.

This could also open the door for additional enhancements, such as history recall from a custom history file specified via another new parameter.

**Comment by iSazonov** [comment]
I think it would be better to implement this in .Net Runtime. I opened an issue for this there.

**Comment by mklement0** [comment]
@iSazonov, you've already opened an issue there? If so, please link to it.

I also wonder what support we need for hosts _other_ than the console (terminal), which would definitely have to be implemented on the PowerShell side.

**Comment by iSazonov** [comment]
The issue https://github.com/dotnet/runtime/issues/800

**Comment by mklement0** [comment]
Thanks, @iSazonov, but I see nothing about the ability to pre-fill (pre-populates) the edit buffer in the linked issue, and the WinAPI's [`ReadConsole`](https://docs.microsoft.com/en-us/windows/console/readconsole) function also doesn't seem to support it,

**Comment by iSazonov** [comment]
@mklement0 The function uses the input buffer as edit buffer and it can be pre-populated.

**Comment by mklement0** [comment]
Thanks, @iSazonov; since I'm not familiar with the  [`ReadConsole`](https://docs.microsoft.com/en-us/windows/console/readconsole) function, let me see if my understanding is correct (happy to hide this comment once we have clarity):

* You're saying that `ReadConsole`'s `lpBuffer` parameter, despite being documented as `_Out_` and only _receiving_ data, can _de facto_ also be used to pre-populate the buffer?

* In other words: the documentation should be amended to note that, right?

**Comment by iSazonov** [comment]
@mklement0 Please look our `internal static string ReadConsole()` method and how it utilizes InitialChars.

**Comment by mklement0** [comment]
Thanks, @iSazonov. I see that you can specify an _offset_ into the buffer, so that `ReadConsole` copies input starting there, but does that mean that the function is still capable of modifying the content _before_ that offset based on user actions (which is crucial for the functionality I'm looking for)? 

(P.S.: There's a grammatical error in one of the variable names: `charsReaded` should be `charsRead`).

**Comment by iSazonov** [comment]
> but does that mean that the function is still capable of modifying the content _before_ that offset based on user actions (which is crucial for the functionality I'm looking for)?

It is my understanding how the code works: we always set the pre-populated buffer.
Also we can edit an input line on Windows :-)

**Comment by eabase** [comment]
@iSazonov 
>  _...we always set the pre-populated buffer. Also we can edit an input line on Windows._

So does this mean that this issue is redundant, fixed, or something else?

**Comment by mklement0** [comment]
@eabase, whatever lower-level features may or may not be _available_ to implement this feature request (I'm still unclear on that part), it hasn't been implemented.

The first step toward that would be for the cmdlets working group to decide on whether the feature is _worth_ implementing.

The issue has been labeled accordingly ("WG-Cmdlets-Utility"), but, unfortunately, there's no telling when this decision will be made.

I _presume_ that the level of community interest helps with prioritizing issues - though I don't know if there's a tool-assisted process for that (such as sorting by the number of thumbs-up votes).

As of this writing, there are 31 feature requests awaiting triage by the cmdlets working group, the oldest dating back to Feb 2021:

https://github.com/PowerShell/PowerShell/issues?q=is%3Aopen+label%3AWG-Cmdlets-Utility+label%3Aneeds-triage+label%3Aissue-enhancement+-label%3Aup-for-grabs+

Note:

* The query explicitly excludes issues that also have an "Up-for-Grabs" label, given that this label implies that an issue _has_ been triaged; such issues should no longer have the "Needs-Triage" label, but I guess there's no automated process to remove the latter. 

* The "In-PR" label by itself does _not_ imply that the issue has been triaged, given that it is automatically assigned when an issue is referenced from a PR and given that the PR is potentially a user's unilateral initiative.

**Comment by mklement0** [comment]
Come to think of it, @iSazonov:

There are many feature requests - this one among them - that do NOT have the "Needs-Triage" label, yet clearly haven't been triaged.

As of this writing, there are 62 such issues (with certain labels that _do_ imply triage having taking place excluded):

https://github.com/PowerShell/PowerShell/issues?page=3&q=is%3Aopen+label%3AWG-Cmdlets-Utility+label%3Aissue-enhancement+-label%3Aneeds-triage+-label%3Aup-for-grabs+-label%3A%22waiting+-+dotnetcore%22++-label%3A%22hacktoberfest-accepted%22

Does this mean that the working groups will _not_ pay attention to untriaged items that lack the "Neeeds-Triage" label, or is the "WG-*" label _alone_ enough (in the absence of  additional labels that imply completed triage)?

If the "Needs-Triage" label is required for the working groups to pay attention to and triage items, then all the issues returned by the query above should be labeled with it.

**Comment by xparadoxon** [comment]
I did a workaround by using Set-Clipboard, maybe this solution works for somebody else too.
`$defaultConnString = "Server=$hostname\$sqlInstance;Database=Demo; User Id=######;Password=######;"`
`Set-Clipboard -Value $defaultConnString`
`$connString = Read-Host "ConnectionString is copied to Clipboard, paste end edit:"`

**Comment by eabase** [comment]
I think this is a great fix that is very simple to implement. It just needs a bump.

**Comment by iSazonov** [comment]
https://github.com/dotnet/runtime/issues/800 was closed with conclusion that `Console.ReadKey()` is recommended 
> Upstack implementations would be wrappers around Console.ReadKey() loops that allow special-case handling of specific keys.

In pwsh port time Windows `ReadConsole` was replaced with Unix `ReadKey()` although it had many limitations in early versions of .Net Core (See https://github.com/dotnet/runtime/pull/72193).
 So sooner or later `ReadConsole` will be replaced by `ReadKey()` and there will be a single code for all platforms.

**Comment by SteveL-MSFT** [comment]
@PowerShell/wg-powershell-cmdlets discussed this.  We agree that adding the ability to have a default value for `Read-Host` makes sense.  However, it would create a technical challenge to print the default value so that the user can then use the cursor to manipulate.  Instead, we propose that if a `-DefaultValue` is specified, the prompt would be rendered as: `<prompt> (default: <defaultValue>): _` where if the user hits `Enter` then the prompt input is the default value, otherwise any characters typed becomes the value.

**Comment by iSazonov** [comment]
> if the user hits `Enter` then the prompt input is the default value, otherwise any characters typed becomes the value.

How could we input empty value?

**Comment by 0x0309** [comment]
> if the user hits `Enter` then the prompt input is the default value, otherwise any characters typed becomes the value.

I thought the whole point was to be able to edit the default without having to re-type the entire value...

**Comment by eabase** [comment]
> > if the user hits `Enter` then the prompt input is the default value, otherwise any characters typed becomes the value.
> 
> How could we input empty value?

That's a good point, and perhaps adding a switch like `-SpaceIsEmpty`?

**Comment by eabase** [comment]
Let's bump again in Sept. 2025?

**Comment by phosjlusky** [comment]
What are people using as a workaround since Read-Host doesn't handle prefilling?

I'm sure there's something better than this:
```powershell
$Value = bash.exe -c "read -e -p 'Enter a value: ' -i 'defaultvalue' ; echo \`$REPLY"
```

**Comment by eabase** [comment]
> Let's bump again in Sept. 2025?

BUMP!

**Comment by phosjlusky** [comment]
Has to be a better way than ``$Value = bash.exe -c "read -e -p 'Enter a value: ' -i 'defaultvalue' ; echo \`$REPLY"``
